keps/sig-api-machinery/2161-apiserver-default-labels/README.md

```diff
@@ -0,0 +1,408 @@
+- open failures with lots of warnings and docs about how old clients might overinterpret the allowed connectivity for a policy
+
+Both of these are impractical due to the nature of security tools, which need to be robust and explicit.
+
+
+Ideally, a fix here would “just work” for “back-rev” implementations of NetworkPolicy (which are out-of-tree - generally implemented by cni plugins) and would not introduce new surprises for users… But in practicality this is not the case because, by adding a new selection data structure, the logical inclusion/exclusion properties of namespace and pod selectors become drastically different, as shown in the table below:
+
+
+
+### Goals
+
+- Add an immutable default label to ALL namespaces, as the namespace name so this can be used as selector by arbitrary components.
+
+### Non-Goals
+
+It’s also worth noting that there are at least 2 other NamespaceSelector APIs in k/k - Validating and Mutating webhooks.  They will inevitably need similar API evolution, but we don't  need to do that as part of this KEP.
+
+## Proposal
+
+We propose to label of all namespaces, by default, with a reserved label (i.e. “kubernetes.io/metadata.namespace” or some such), to allow easy selection.
+If this label is missing, it is added by the apiserver
+If the label is incorrect or changed, the apiserver fails to validate the object
+
+
+### User Stories (Optional)
+
+See https://github.com/kubernetes/enhancements/pull/2113/files for relevant user stories.
+
+### Notes/Constraints/Caveats (Optional)
+
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+
+This can be implemented by modifying the defaults.go and validation.go components for the namespace API, i.e., in the defaults.go file for api/core:
+
+```
+func SetDefaults_Namespace(obj *v1.Namespace) {
+if obj.Labels == nil {
+   		obj.Labels = map[string]string{}
+}
+if _, found := obj.Labels[“kubernetes.io/metadata.name”]; !found {
+   		obj.Labels[“kubernetes.io/metadata.name”] = obj.Name
+}
+}
+```
+And in the validation.go, we implement the guarantee of this namespace’s value being constant WRT namespace name (which is immutable).
+
+```
+func ValidateNamespace(namespace *core.Namespace) field.ErrorList {
+	// ...
+   	// Check if namespace.Labels[“kubernetes.io/metadata.name”] == namespace.Name
+// if not add an error
+   	// ...
+return allErrs
+```
+
+This solves not just NetworkPolicy, but all such namespace selector APIs without any disruption or version skew problems.
+
+### Test Plan
+
+### Graduation Criteria
+
+
+### Upgrade / Downgrade Strategy
+
```

Comment:
Need to check the watch issue above.  If that works, great.  If not, we may need to add a sweeper that populates the label at startup.

Reply:
In case of downgrade, any policies that rely on this will either work fine (if the label was written) or fail-closed becasuse the label does not exist.  Acceptable.

Reply:
Since reads go through `defaults.go` you get lazy population in watches, so this works as intended